README.md

```diff
@@ -10,46 +10,167 @@ Stage: 1
-to be in the form of ECMAScript syntax that enables the module import syntax to
-include import reflection syntax to indicate an alternative reflective API import,
-with the primary use case in mind of permitting these module type imports for Wasm.
-
-[Module Linking Proposal]: https://github.com/WebAssembly/module-linking/blob/master/proposals/module-linking/Explainer.md
-[wasm-esm]: https://github.com/WebAssembly/esm-integration/tree/master/proposals/esm-integration
+For both JavaScript and WebAssembly, there is a need to be able to more closely
+customize the loading, linking and execution of modules beyond the standard host
+execution model.
+
+For WebAssembly, imports and exports for WebAssembly models often require custom
+inspection and wrapping in order to behave correctly, which can be better
+provided via manual instantiation than relying directly on the base
+[ESM integration][wasm-esm] proposal.
+
+For JavaScript, creating userland loaders would require a module reflection type
+in order to share the host parsing, execution, security and caching semantics.
 
 ## Proposal
 
-The proposal is to permit an optional string literal attribute to be associated
-with any import:
+This proposal enables reflection capabilities for both JavaScript and
+WebAssembly modules by enabling a new type of import, an import reflection:
 
 ```js
 import x from "<specifier>" as "<reflection-type>";
 ```
 
-Here the import reflection type is added to the end of the ImportStatement,
+The type of the reflection object for WebAssembly would be a
+`WebAssembly.Module`, as defined in the
+[WebAssembly JS integration API][wasm-js-api].
+
+The reflection would represent an unlinked and uninstantiated module, while
+still being able to support the same CSP policy as the native
+[ESM integration][wasm-esm], avoiding the need for `unsafe-wasm-eval` for
+custom Wasm execution.
+
+Type type of the reflection object for a JavaScript module would be defined
+explicitly in this specification as a `SourceTextModule` class with the
+following interface:
+
+```js
+class SourceTextModule {
+  // create a new instance of this module, with the given global environment record
+  instantiate (globalEnvironmentRecord: Record<string, Binding>): ModuleInstance;
+  
+  // static function to retrieved the named exports, where "module" corresponds
+  // to the module from which this export is reexported (if any) and "name"
+  // corresponds to the exported name (or * for the case of a star re-export).
+  static exports(): { module?: string, name: string | '*' }[];
+
+  // static function to retrieve imported binding list, where "module"
+  // corresponds to the module specifier being imported and "name" corresponds
+  // to the exported binding from the imported module.
+  static imports(): { module: string, name: string }[];
+}
+```
+
+`SourceTextModule.exports(module)` and `SourceTextModule.imports(module)` can
```

Comment:
It may well make sense for `SourceTextModule` to be a global, especially if it were to have a string constructor userland version for custom module evaluation.

Reply:
I've explicitly noted `SourceTextModule` may be a global.